ansible/verify_config.py

```diff
@@ -0,0 +1,768 @@
+    TESTBED_FILE = "testbed.yaml"
+    TOPO_FILE_PATTERN = "vars/topo*.yml"
+    VM_FILE = "veos"
+    FANOUT_LINKS_FILE = "files/sonic_*_links.csv"
+    FANOUT_DEVICES_FILE = "files/sonic_*_devices.csv"
+    FANOUT_BMC_LINKS_FILE = "files/sonic_*_bmc_links.csv"
+    FANOUT_PDU_LINKS_FILE = "files/sonic_*_pdu_links.csv"
+    FANOUT_CONSOLE_LINKS_FILE = "files/sonic_*_console_links.csv"
+    FANOUT_GRAPH_GROUP_FILE = "files/graph_groups.yml"
+    GROUP_VARS_ENV_FILE = "group_vars/all/env.yml"
+
+
+class Utility:
+    @staticmethod
+    @lru_cache
+    def parse_yml(file):
+        with open(file, "r") as stream:
+            return yaml.safe_load(stream)
+
+    @staticmethod
+    @lru_cache
+    def get_devices_from_links_file(file):
+        devices = set()
+
+        try:
+            with open(file, "r") as stream:
+                for row in csv.DictReader(stream):
+                    devices.add(row['StartDevice'])
+                    devices.add(row['EndDevice'])
+        except FileNotFoundError:
+            log.error(f"Cannot find file {file} while getting devices information")
+
+        return devices
+
+    @staticmethod
+    @lru_cache
+    def get_devices_from_devices_file(file):
+        devices = set()
+
+        try:
+            with open(file, "r") as stream:
+                for row in csv.DictReader(stream):
+                    devices.add(row['Hostname'])
+        except FileNotFoundError:
+            log.error(f"Cannot find file {file} while getting devices information")
+
+        return devices
+
+    @staticmethod
+    @lru_cache
+    def get_topo_from_var_files():
+        topo_name_set = set()
+
+        for topo_file_path in glob.glob(os.path.abspath(Config.TOPO_FILE_PATTERN)):
+            file_name, _ = os.path.basename(topo_file_path).split(".")
+            topo_name = file_name[len("topo_"):]
+            topo_name_set.add(topo_name)
+
+        return topo_name_set
+
+    @staticmethod
+    @lru_cache
+    def get_inv_name_from_file(link_file):
+        inv_name_set = set()
+
+        for inv_file_path in glob.glob(os.path.abspath(link_file)):
+            file_name, _ = os.path.basename(inv_file_path).split(".")
+
+            inv_name = file_name[len("sonic_"): file_name.index("_", len("sonic_"))]
+            inv_name_set.add(inv_name)
+
+        return inv_name_set
+
+    @staticmethod
+    @lru_cache
+    def get_num_vm(topo_name):
+        topology = Utility.parse_yml(Config.TOPO_FILE_PATTERN.replace("*", f"_{topo_name}"))
+        if 'topology' not in topology or 'VMs' not in topology['topology']:
+            return 0
+
+        return len(topology['topology']['VMs'])
+
+    @staticmethod
+    def execute_shell(cmd, is_shell=False):
+        try:
+            return subprocess.run(
+                cmd.split(" "),
+                shell=is_shell,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE
+            ).returncode
+        except FileNotFoundError:
```

Comment:
I found this interesting as well. When running a command that is not in `PATH`, `subprocess.run()` will throw `FileNotFoundError`. I assume it's because fundamentally, a command is just a file in `PATH` and `subprocess.run()` is trying to find that file to execute.

For example, run this in your python interpreter would throw the same error.

```python
import subprocess
subprocess.run(["randomcommand"])
```

I'm open for suggestions here to see how we could improve this

Reply:
Ah I see. Given that we are only interested in the return code of the command, I think your implementation is good. Thanks!